simulation/chart.py

```diff
@@ -1,232 +1,438 @@
 
-    if size is not None:
-        config["symbolSize"] = size
+class PenConfig(TypedDict, total=False):
+    """Configuration for pen to draw lines"""
 
-    if color is not None:
-        config["symbolBrush"] = color
+    width: float
+    """Width of the line"""
+    style: PenStyle
+    """Line style"""
+    color: Color
+    """Line color"""
 
-    if symbol is not None:
-        config["symbol"] = symbol
 
-    if border is not None:
-        config["symbolPen"] = border
+# Method to resample the data before plotting ot avoid plotting multiple line segments per pixel.
+DownsampleMethod = Literal[
+    "subsample",  # Downsample by taking the first of N samples. This method is fastest and least accurate.
+    "mean",  # Downsample by taking the mean of N samples.
+    "peak",  # Downsample by drawing a saw wave that follows the min and max of the original data. This method produces the best visual representation of the data but is slower.
+]
 
-    return chart.plot(name=name, pen=None, **config)
+DEFAULT_WIN_SIZE = (1000, 1000)
+DEFAULT_CODEC = "libx264"
+DEFAULT_PIXEL_FORMAT = "yuv420p"
 
 
-def create_video_process(
-    file_name: str,
-    size: Tuple[int, int],
-    codec: str = "libx264",
-    pixel_format: str = "yuv420p",
-) -> Popen:
-    """Creates a ffmpeg video process which accepts input from stdin.(NA in windows)
+class ChartWidget:
+    """A chart container for single group of plots and a legend.
 
     Args:
-        file_name (str): name of the file to save encoded video ouput
-        size (Tuple[int,int]): Size (width,height) of the video frame
-        codec (str, optional): Codec name to be used for encoding. Defaults to "libx264".
-        pixel_format (str, optional): Ouput pixel format. Defaults to "yuv420p".
-
-    Returns:
-        Process: Asyncronious sub process listening stdin for encoding to the file.
+        chart (PlotItem): base container plotitem
     """
-    process = (
-        ffmpeg.input(
-            "pipe:",
-            format="rawvideo",
-            pix_fmt="rgb24",
-            s=f"{size[0]}x{size[1]}",
+
+    def __init__(self, chart: pg.PlotItem) -> None:
+        self.__chart = chart
+        self.__plots: Dict[str, pg.PlotDataItem] = {}
+
+    @staticmethod
+    def _update_downsapling(
+        plot: pg.PlotDataItem,
+        auto: Optional[bool] = None,
+        ds: Optional[int] = None,
+        method: Optional[DownsampleMethod] = None,
+    ):
+        """Update plot downsampling parameters"""
+
+        if auto is None and ds is None and method is None:
+            return
+
+        config: Dict[str, Any] = {}
+
+        if auto is not None:
+            config["auto"] = auto
+
+        if ds is not None:
+            config["ds"] = ds
+
+        if method is not None:
+            config["method"] = method
+
+        plot.setDownsampling(**config)
+
+    def add_line_plot(
+        self,
+        id: str,
+        name: str,
+        color: Optional[Color] = None,
```

Comment:
I'm not a massive fan of using `None`. See Tony Hoare's (the guy who invented `None`) [talk](https://www.infoq.com/presentations/Null-References-The-Billion-Dollar-Mistake-Tony-Hoare/) on why this is something he regrets introducing to the world. Can we give sensible defaults here?

Reply:
Actually yes. I can pop hidden defaults from pyqtgraph method. see `self.opts` in `PlotDataItem` constructor.

Reply:
Cool. Let's use those defaults then. Thanks!

Reply:
I extracted as much as I can, and retained some where None has a meaning such as non existent title for a plot axis.
https://github.com/semiotic-ai/autoagora-agents/pull/21/commits/d6582cf6a8d6916239e09a4f62d266c848672a60